### DISCUSSION TITLE
Does @itwin/desktop-viewer-react support mapLayerOptions?

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#5647] [Q&A] [answered]
I was able to successfully download a copy of the sample **3d-viewer** and display the background using my keys in mapLayerOptions.

However I was not able to do the same thing with a desktop viewer that I created using:

`npx create-react-app 3d-viewer-desktop --template @itwin/desktop-viewer --scripts-version @bentley/react-scripts`

From the electron app debug Network tab I see the following error:

![image](https://github.com/iTwin/itwinjs-core/assets/30668739/a852b810-3a46-4667-9cc0-5adbc83feaef)

**Comment by aruniverse** [comment, marked as answer]
Hi john,

Could you link to the `3d viewer` you reference?

Yes you can provide map-layer-options to your desktop viewer like so:
<img width="662" alt="image" src="https://github.com/iTwin/itwinjs-core/assets/11051042/6ba096e6-1860-4752-b22a-1dd0812a800a">

```tsx
  const desktopInitializerProps = useMemo<DesktopInitializerParams>(
    () => ({
      clientId: process.env.IMJS_VIEWER_CLIENT_ID ?? "",
      rpcInterfaces: viewerRpcs,
      additionalI18nNamespaces: ["iTwinDesktopViewer"],
      enablePerformanceMonitors: true,
      onIModelAppInit,
      mapLayerOptions: {
        BingMaps: {
          key: "key",
          value: process.env.IMJS_BING_MAPS_KEY ?? "",
        },
        MapboxImagery: {
          key: "access_token",
          value: process.env.IMJS_MAPBOX_ACCESS_TOKEN ?? "",
        },
      },
    }),
    [onIModelAppInit]
  );
```

**Comment by JohnMajerle** [comment]
Hello Arun,

This is the web viewer sample I referred to that works OK:
[3D Viewer](https://www.itwinjs.org/sandboxes/iTwinPlatform/3d%20Viewer)

If  you create an Electron desktop app using;

`npx create-react-app 3d-viewer-desktop --template @itwin/desktop-viewer --scripts-version @bentley/react-scripts`

...the viewer portion is inside ViewerRoute.tsx which I have pasted below.  The only line I added was:

`mapLayerOptions={mapLayerOptions}`

The variable mapLayerOptions is from MapLayerOptions.ts that I used successfully in the web viewer so I am confident that the keys are OK.

Note that with my desktop viewer I am opening a local bim file and have not signed into CONNECT.

It should only take a few minutes for someone  to reproduce  these steps to verify if background display works  with an application created from the @itwin/desktop-viewer template.

Thanks,

John M.




**ViewerRoute.tsx**

*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import {
  Viewer,
  ViewerContentToolsProvider,
  ViewerNavigationToolsProvider,
  ViewerStatusbarItemsProvider,
} from "@itwin/desktop-viewer-react";
import { MeasureToolsUiItemsProvider } from "@itwin/measure-tools-react";
import { PropertyGridUiItemsProvider } from "@itwin/property-grid-react";
import { TreeWidgetUiItemsProvider } from "@itwin/tree-widget-react";
import React, { useEffect, useState } from "react";
import { useLocation } from "react-router-dom";

import { viewerRpcs } from "../../../common/ViewerConfig";
import { IModelMergeItemsProvider } from "../../extensions";

import { mapLayerOptions, tileAdminOptions } from "./MapLayerOptions"

export interface ViewerRouteState {
  filePath?: string;
}

export const ViewerRoute = () => {
  const location = useLocation();
  const [filePath, setFilePath] = useState<string>();

  useEffect(() => {
    const routeState = location?.state as ViewerRouteState | undefined;
    if (routeState?.filePath) {
      setFilePath(routeState?.filePath);
    }
  }, [location?.state]);

  return filePath ? (
    <Viewer
      clientId={process.env.IMJS_VIEWER_CLIENT_ID ?? ""}
      rpcInterfaces={viewerRpcs}
      filePath={filePath}
      uiProviders={[
        new ViewerNavigationToolsProvider(),
        new ViewerContentToolsProvider({
          vertical: {
            measureGroup: false,
          },
        }),
        new ViewerStatusbarItemsProvider(),
        new TreeWidgetUiItemsProvider(),
        new PropertyGridUiItemsProvider({
          enableCopyingPropertyText: true,
        }),
        new MeasureToolsUiItemsProvider(),
        new IModelMergeItemsProvider(),
      ]}
      enablePerformanceMonitors={true}
      mapLayerOptions={mapLayerOptions}
    />
  ) : null;
};

**Comment by JohnMajerle** [comment]
@aruniverse  I'm still waiting on an answer to this question.  Have you been able to get background graphics to display using an Electron app created from the template @itwin/desktop-viewer and opening a local bim file without sign in?

**Reply by aruniverse** [reply]
Hi john, as mentioned in [this comment]( https://github.com/iTwin/itwinjs-core/discussions/5647#discussioncomment-6189538), you need to supply the mapLayersOptions in the `useDesktopInitializer` hook in `App.tsx` not in `ViewerRoute.tsx` like you mentioned in your last comment

**Comment by JohnMajerle** [comment]
@aruniverse 
Hello Arun,

OK, that worked.  But I am confused why this is necessary since mayLayerOptions is still a property of DesktopViewerProps which is used to create a Viewer object from @itwin/desktop-viewer-react.

**Reply by aruniverse** [reply]
The desktop's Viewer component actually passes some of the DesktopViewerProps to the `useDesktopinitializer` hook, but  in the desktop viewer template we call that hook before the Viewer component gets rendered, so things will already be initialized and passing in mapLayersOptions afterwards will have no effect
